pkg/hive/hive.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package hive
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/hive/pb"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/p2p/protobuf"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+const (
+	protocolName    = "hive"
+	protocolVersion = "1.0.0"
+	peersStreamName = "peers"
+	maxPO           = 7
+	messageTimeout  = 5 * time.Second // maximum allowed time for a message to be read or written.
+)
+
+type Service struct {
+	streamer          p2p.Streamer
+	connectionManager ConnectionManager
+	peerSuggester     PeerSuggester
+	addressFinder     AddressFinder
+	logger            logging.Logger
+}
+
+type Options struct {
+	Streamer          p2p.Streamer
+	ConnectionManager ConnectionManager
+	PeerSuggester     PeerSuggester
+	AddressFinder     AddressFinder
+	Logger            logging.Logger
+}
+
+func New(o Options) *Service {
+	return &Service{
+		streamer:          o.Streamer,
+		logger:            o.Logger,
+		connectionManager: o.ConnectionManager,
+		peerSuggester:     o.PeerSuggester,
+		addressFinder:     o.AddressFinder,
+	}
+}
+
+type BzzAddress struct {
+	Overlay, Underlay []byte
+}
+
+type ConnectionManager interface {
+	// todo: this can be the libp2p.Connect or something else
+	Connect(underlay []byte) error
+}
+
+type PeerSuggester interface {
+	SuggestPeers(peer p2p.Peer, bin, limit int) (peers []p2p.Peer)
+}
+
+type AddressFinder interface {
+	Underlay(overlay swarm.Address) (underlay []byte, err error)
+}
+
+func (s *Service) Protocol() p2p.ProtocolSpec {
+	return p2p.ProtocolSpec{
+		Name:    protocolName,
+		Version: protocolVersion,
+		Init:    s.Init,
+		StreamSpecs: []p2p.StreamSpec{
+			{
+				Name:    peersStreamName,
+				Handler: s.peersHandler,
+			},
+		},
+	}
+}
+
+// Init is called when the new peer is being initialized.
+// This should happen after overlay handshake is finished.
+func (s *Service) Init(peer p2p.Peer) error {
```

Comment:
Should we accept context here and pass it to getPeers?

Reply:
Sure! Did not think about context yet, but maybe all methods called from libp2p should have context?